platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/migratable/TaskList.java

```diff
@@ -107,16 +119,46 @@ public void add(Operation op, String id, Predicate<ProfileMigrationReport> task)
   }
 
   /**
-   * Increases the number of attempts left for the specified operation
+   * Adds a new or retrieves an already registered compound task to be executed for a specific
+   * operation. This method expects to be called consistently for each subtasks of a given operation
+   * as the container and the compound task will only be created and registered the first time this
+   * method is called. The returned {@link CompoundTask} object provides a way for the client to add
+   * subtask information to the container that will later be provided to the registered compound
+   * task when it is executed.
+   *
+   * <p><i>Note:</i> It is recommended to call this method for a given operation in one single place
+   * in your code to ensure consistency of the container type and the compound task for each
+   * subtasks being added.
+   *
+   * <p>
    *
-   * @param op the operation to increase the number of attempts for
-   * @throws IllegalArgumentException if <code>op</code> is <code>null</code>
+   * <pre><code>
+   *   private final TaskList tasks = new TaskList();
+   *   ...
+   *   tasks.addIfAbsent(
+   *     Operation.INSTALL,
+   *     HashSet<MyObjectClass>::new,
+   *     (objects, r) -> [do something with the accumulated objects]
+   *   ).add(name, HashSet::add);
+   * </code></pre>
+   *
+   * @param <T> the type of the container used to accumulate recorded subtasks
+   * @param op the operation for which to record the subtask
+   * @param containerFactory a supplier which returns a new, empty container object (only called the
+   *     first time this method is called)
+   * @param task the compound task to add for later execution (the task will only be registered the
+   *     first time this method is called for a given operation)
+   * @throws IllegalArgumentException if <code>op</code>, <code>containerFactory</code>, or <code>
+   *     task</code>, is <code>null</code>
    */
-  public void increaseAttemptsFor(Operation op) {
+  public <T> CompoundTask<T> addIfAbsent(
```

Comment:
❗️ I think that passing the `task` along with the `Operation` to `addIfAbsent` makes things more confusing. Which one needs to be absent before it's added, the `Operation` or the `task`? Also, the `Supplier` and `accumulator` are closely related, so they really make more sense being provided at the same time and not in separate method calls. I think the original approach was better.
❓ Could we leverage `Collectors` here (didn't try that code so it might need some tweaking)?
```
public class TaskList {
  public <T> CompountTask<T> addIfAbsent(operation,
      Collector<CompoundTask<T>, ?,  R> compoundTaskCollector) {
    // ...
  }
}

// Usage

addIfAbsent(operation, Collectors.toSet()).add(task);

```

Reply:
Based on our discussion, I will leave it as is.

But quickly, I cannot use the collector paradigm here since the CompoundTask object acts as a container where the client adds data to it as opposed to a stream where the stream is responsible for providing the data. I am forced to use an accumulator function since I do not know inside the CompoundTask what is the type of container used so I cannot simply take whatever object provided on the API and add it to that container. I also cannot provide access to that container for the user to add directly to it as the TaskList implementation needs to track if subtasks were added or not to a compound task.

As for the rest, the `addIfAbsent()` registers a compound task for a given operation to which one can add subtasks (represented with whatever data the user intends to track) multiple time. So the accumulation (or addition to the container) is done on the compound task which needs only to be created once which is why the method on the TaskList is now called `addIfAbsent()` (kinda like the `computeIfAbsent()` of maps.